src/edu/washington/escience/myria/api/encoding/QueryConstruct.java

```diff
@@ -495,61 +504,47 @@ private static RootOperator instantiateFragment(final PlanFragmentEncoding planF
         fragmentRoot = (RootOperator) op;
       }
     }
-    allOperators.putAll(myOperators);
-
-    for (IDBControllerEncoding idb : idbs) {
-      nonIterativeConsumers.remove(idb.argIterationInput);
-      nonIterativeConsumers.remove(idb.argEosControllerInput);
-    }
-
-    Set<PlanFragmentEncoding> dependantFragments = Sets.newHashSet();
-    for (AbstractConsumerEncoding<?> c : nonIterativeConsumers.values()) {
-      dependantFragments.add(opOwnerFragment.get(c.argOperatorId));
-    }
-
-    for (PlanFragmentEncoding f : dependantFragments) {
-      instantiateFragment(f, args, instantiatedFragments, opOwnerFragment, allOperators);
-    }
-
-    for (AbstractConsumerEncoding<?> c : nonIterativeConsumers.values()) {
-      Consumer consumer = (Consumer) myOperators.get(c.opId);
-      Integer producingOpName = c.argOperatorId;
-      Operator producingOp = allOperators.get(producingOpName);
-      if (producingOp instanceof IDBController) {
-        consumer.setSchema(IDBController.EOI_REPORT_SCHEMA);
-      } else {
-        consumer.setSchema(producingOp.getSchema());
-      }
-    }
-
-    /* Connect all the operators. */
-    for (OperatorEncoding<?> encoding : planFragment.operators) {
-      encoding.connect(myOperators.get(encoding.opId), myOperators);
-    }
-
-    for (IDBControllerEncoding idb : idbs) {
-      IDBController idbOp = (IDBController) myOperators.get(idb.opId);
-      Consumer eosControllerInput = (Consumer) idbOp.getChildren()[IDBController.CHILDREN_IDX_EOS_CONTROLLER_INPUT];
-      eosControllerInput.setSchema(EOSController.EOS_REPORT_SCHEMA);
-      Operator iterativeInput = idbOp.getChildren()[IDBController.CHILDREN_IDX_ITERATION_INPUT];
-      if (iterativeInput instanceof Consumer) {
-        Operator initialInput = idbOp.getChildren()[IDBController.CHILDREN_IDX_INITIAL_IDB_INPUT];
-        ((Consumer) iterativeInput).setSchema(initialInput.getSchema());
-        // Note: better to check if the producer of this iterativeInput also has the same schema, but can't find a good
-        // place to add the check given the current framework.
-      }
-    }
-
     if (fragmentRoot == null) {
       throw new MyriaApiException(Status.BAD_REQUEST, "No " + RootOperator.class.getSimpleName()
           + " detected in the fragment.");
     }
-
-    instantiatedFragments.put(planFragment, fragmentRoot);
     return fragmentRoot;
   }
 
   /**
+   * 
+   * @param fragments the JSON-encoded query fragments to be executed in parallel
+   * @param allOperators a map to keep instantiated operators.
+   * @return if all the consumers have their schemata set.
+   */
+  private static boolean setConsumerSchema(final List<PlanFragmentEncoding> fragments,
```

Comment:
Yes recursive propagation is possible, e.g.
Consumer_B -> IDBController -> Producer_A
Consumer_A -> IDBController -> Producer_B
If we need to know the schema of Consumer_A, we need the schema of Producer_A then Consumer_B then Producer_B. In this case since an IDBController has two children and one is its initial input which usually has a fixed schema, we won't have infinite loop, but you're right that in general it's possible to construct a plan with infinite loop. I'll address it in the following commits.